uefi/src/mem.rs

```diff
@@ -0,0 +1,134 @@
+//! This is a utility module with helper methods for allocations/memory.
+
+use crate::ResultExt;
+use crate::{Result, Status};
+use ::alloc::{alloc, boxed::Box};
+use core::alloc::Layout;
+use core::fmt::Debug;
+use core::ptr::NonNull;
+use core::slice;
+use uefi::data_types::Align;
+
+/// Helper to return owned versions of certain UEFI data structures on the heap, hence, in a
+/// [`Box`]. It is intended to wrap low-level UEFI functions off this crate that
+/// - consume a mutable reference to a buffer that needs to be filled with some data of type,
+/// - return a mutable typed reference (which points to the same memory as the input buffer) on
+///   success, and
+/// - return the required buffer size on error.
+pub fn make_boxed<
+    'a,
+    Data: Align + ?Sized + Debug + 'a,
+    F: FnMut(&'a mut [u8]) -> Result<&'a mut Data, Option<usize>>,
+>(
+    mut fetch_data_fn: F,
+) -> Result<Box<Data>> {
+    let required_size: usize = match fetch_data_fn(&mut [])
+        .expect_err("succeeded unexpectedly")
+        .split()
+    {
+        (s, None) => Err::<usize, crate::result::Error>(s.into()),
+        (_, Some(required_size)) => Ok(required_size),
+    }
+    .unwrap();
```

Comment:
Suggestion:
```rust
    let required_size = match fetch_data_fn(&mut []).map_err(Error::split) {
        // This is the expected case: the empty buffer passed in is too
        // small, so we get the required size.
        Err((Status::BUFFER_TOO_SMALL, Some(required_size))) => Ok(required_size),
        // Propagate any other error.
        Err((status, _)) => Err(Error::from(status)),
        // Success is unexpected, return an error.
        Ok(_) => Err(Error::from(Status::UNSUPPORTED)),
    }?;
```

This avoids a couple unwraps, propagating errors instead, and explicitly checks for the pair of `BUFFER_TOO_SMALL` and `Some(required_size)` so that we know we're getting exactly what we expect.

Reply:
thanks for the suggestion, I like it. Looks much cleaner!